sway/input/seat.c

```diff
@@ -1091,23 +1093,33 @@ void seat_set_focus(struct sway_seat *seat, struct sway_node *node) {
 		node->sway_container : NULL;
 
 	// Deny setting focus to a view which is hidden by a fullscreen container
+	// except if forced
```

Comment:
Design question: could we refactor out this code into a new `struct sway_container *container_obstructing_fullscreen_container(struct sway_container *)`?

Then `focus.c` could check if `container_obstructing_fullscreen_container(...) != NULL` and `container_fullscreen_disable(...)` that, removing the need to plumb `force` through this code.

Reply:
This sounds like a much better design, thank you.
I've made the changes accordingly.

Reply:
:+1: Ideally, we'd use this code within `seat_set_focus` too, so we don't have the same block duplicated twice.

Reply:
Which block are you talking about? The one added in focus.c?

Initially, that's why I added a `seat_set_focus_force` function as when I reviewed the code back in January it seemed like this was the only location we'd actually need to do that. There are a lot of calls to `seat_set_focus*` and adding the same logic there has a big chance of introducing regressions IMO.

Reply:
I meant that these two blocks are basically identical in functionality:

https://github.com/swaywm/sway/blob/63564ee11dd168e35b7e1ac7f3068fff10372cbd/sway/tree/container.c#L420-L441

https://github.com/swaywm/sway/blob/63564ee11dd168e35b7e1ac7f3068fff10372cbd/sway/input/seat.c#L1093-L1112

The one from `seat.c` can be made to call `container_obstructing_fullscreen_container` and avoid duplicating its logic.

Reply:
Ah gotcha haha :smiley: !